src/mds/CDir.cc

```diff
@@ -2651,9 +2651,26 @@ void CDir::set_dir_auth(const mds_authority_t &a)
   bool was_subtree = is_subtree_root();
   bool was_ambiguous = dir_auth.second >= 0;
 
+  mds_authority_t old_auth = dir_auth;
   // set it.
   dir_auth = a;
 
+  if (old_auth != dir_auth) {
+    if (dir_auth != CDIR_AUTH_DEFAULT) {
```

Comment:
if "old_auth != CDIR_AUTH_DEFAULT", need remove old one from subtree_roots.   After switching to counter, please add code that updates the counter to code blocks "if (!was_subtree && is_subtree_root())" and "if (was_subtree && !is_subtree_root())"